Unsupported situations for loading "Sum" layer using readNetFromTensorflow

<!--
If you have a question rather than reporting a bug please go to https://forum.opencv.org where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

##### System information (version)

- OpenCV => 4.5.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2019


##### Detailed description

<!-- your description -->
I am trying to use function "readNetFromTensorflow()" in dnn module to load a pb model.  But I got the error.
![SY6M5JX$ NVR91)T)A B$%L](https://user-images.githubusercontent.com/56301098/107753310-e1572100-6d5a-11eb-9404-4e19c81d1be8.png)
The structure around  is import/layer2/sum is
![~DH 4(GU)`J4`OS{$}6BJ}Q](https://user-images.githubusercontent.com/56301098/107753429-051a6700-6d5b-11eb-8c01-ed36cec80cba.png)
I read the source code and found when the parameter "reduction_indices" is -1 or 3, the "Sum" layer won't be added to "layer_id". So its next layer can't find the corresponding input layer. Is that mean opencv/dnn dose not support for loading such "Sum" layer ? 


Can anyone help me ?

 
##### Steps to reproduce
The full Code is 


```
#include <fstream>
#include <sstream>
#include <opencv2/dnn.hpp>

using namespace cv;
using namespace dnn;
int main(int argc, char** argv)
{
	String modelPath = "3dssdOptimize1.pb";
	Net net = readNetFromTensorflow(modelPath);
	std::vector<cv::String> layerNames = net.getLayerNames();
	return 0;
}
```


##### Issue submission checklist

 - [x] I report the issue, it's not a question
   <!--
   OpenCV team works with forum.opencv.org, Stack Overflow and other communities
   to discuss problems. Tickets with question without real issue statement will be
   closed.
   -->
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   <!--
   Places to check:
   * OpenCV documentation: https://docs.opencv.org
   * FAQ page: https://github.com/opencv/opencv/wiki/FAQ
   * OpenCV forum: https://forum.opencv.org
   * OpenCV issue tracker: https://github.com/opencv/opencv/issues?q=is%3Aissue
   * Stack Overflow branch: https://stackoverflow.com/questions/tagged/opencv
   -->
 - [x] I updated to latest OpenCV version and the issue is still there
   <!--
   master branch for OpenCV 4.x and 3.4 branch for OpenCV 3.x releases.
   OpenCV team supports only latest release for each branch.
   The ticket is closed, if the problem is not reproduced with modern version.
   -->
 - [ ] There is reproducer code and related data files: videos, images, onnx, etc
   <!--
   The best reproducer -- test case for OpenCV that we can add to the library.
   Recommendations for media files and binary files:
   * Try to reproduce the issue with images and videos in opencv_extra repository
     to reduce attachment size
   * Use PNG for images, if you report some CV related bug, but not image reader
     issue
   * Attach the image as archive to the ticket, if you report some reader issue.
     Image hosting services compress images and it breaks the repro code.
   * Provide ONNX file for some public model or ONNX file with with random weights,
     if you report ONNX parsing or handling issue. Architecture details diagram
     from netron tool can be very useful too. See https://lutzroeder.github.io/netron/
   -->

Thank you for update!

Debug log messages are below:

```
...
[DEBUG:0] global dnn/src/tensorflow/tf_importer.cpp (764) predictOutputDataLayout DNN/TF: predictOutputDataLayout(import/layer2/Square @ Square) => 0 (from inputs)
[DEBUG:0] global dnn/src/tensorflow/tf_importer.cpp (884) populateNet DNN/TF: (452/847) Parse layer import/layer2/Sum @ Sum with 2 inputs
[DEBUG:0] global dnn/src/tensorflow/tf_importer.cpp (764) predictOutputDataLayout DNN/TF: predictOutputDataLayout(import/layer2/Sum @ Sum) => 0 (from inputs)
[DEBUG:0] global dnn/src/tensorflow/tf_importer.cpp (884) populateNet DNN/TF: (453/847) Parse layer import/layer2/add @ Add with 2 inputs
[DEBUG:0] global dnn/src/tensorflow/tf_importer.cpp (764) predictOutputDataLayout DNN/TF: predictOutputDataLayout(import/layer2/add @ Add) => 0 (from inputs)
[ERROR:0] global dnn/src/tensorflow/tf_importer.cpp (2477) parseNode DNN/TF: Can't parse layer for node='import/layer2/add'. Exception: OpenCV(4.5.1-dev) dnn/src/tensorflow/tf_importer.cpp:1184: error: (-2:Unspecified error) Input layer not found: import/layer2/Sum in function 'parseNode'

unknown file: Failure
C++ exception with description "OpenCV(4.5.1-dev) dnn/src/tensorflow/tf_importer.cpp:1184: error: (-2:Unspecified error) Input layer not found: import/layer2/Sum in function 'parseNode'
" thrown in the test body.
```

"Missing" layer is processed on step 452, so it should exist.

After correction of the case, where ``reduction_indices=-1 or 3`` there are some more issues when feeding ``3dssdOptimize1.pb`` model into OCV:
1. problem in ``import/layer3/add_2`` node:
 ```
.../opencv/modules/dnn/src/tensorflow/tf_importer.cpp:1179: error: (-215:Assertion failed) values.type() == CV_32FC1 in function 'parseNode'
```
``import/layer3/add_2/y`` const has ``int32`` type:

![issue1](https://user-images.githubusercontent.com/7347318/112856678-b7627f80-90b8-11eb-85ee-57494e1ee1a8.png)
However, [tf_importer.cpp](https://github.com/opencv/opencv/blob/master/modules/dnn/src/tensorflow/tf_importer.cpp#L1179) expects ``CV_32FC1`` type. 
@alalek, could you, please, clarify, could we provide cast to ``CV_32FC1`` in such cases without assertion or it would cause accuracy or some other related issues?

2. If the type problem is skipped, we will get the following:
```
.../opencv/modules/dnn/src/tensorflow/tf_importer.cpp:616: error: (-2:Unspecified error) Input layer not found: import/zeros_like in function 'connect'
```
Unfolded ``import/zeros_like`` const was ignored:

![issue2](https://user-images.githubusercontent.com/7347318/112869611-082ca500-90c6-11eb-8e44-963928af32b2.png)
Hence, in ``import/stack`` node this node will not be found. It could be solved with [text graph generation](https://github.com/opencv/opencv/wiki/TensorFlow-text-graphs) or some model optimizations.

--- a/modules/dnn/src/tensorflow/tf_importer.cpp
+++ b/modules/dnn/src/tensorflow/tf_importer.cpp
@@ -2360,12 +2360,9 @@ void TFImporter::parseNode(const tensorflow::NodeDef& layer_)
                         // To keep correct order after squeeze dims we first need to change layout from NCHW to NHWC
                         LayerParams permLP;
                         int order[] = {0, 2, 3, 1};  // From OpenCV's NCHW to NHWC.
-                        permLP.set("order", DictValue::arrayInt<int*>(order, 4));
                         std::string permName = name + "/nchw";
-                        CV_Assert(layer_id.find(permName) == layer_id.end());
-                        int permId = dstNet.addLayer(permName, "Permute", permLP);
-                        layer_id[permName] = permId;
-                        connect(layer_id, dstNet, Pin(name), permId, 0);
+                        Pin inpId = Pin(name);
+                        addPermuteLayer(order, permName, inpId);
 
                         LayerParams squeezeLp;
                         std::string squeezeName = name + "/squeeze";
@@ -2375,6 +2372,38 @@ void TFImporter::parseNode(const tensorflow::NodeDef& layer_)
                         int squeezeId = dstNet.addLayer(squeezeName, "Flatten", squeezeLp);
                         layer_id[squeezeName] = squeezeId;
                         connect(layer_id, dstNet, Pin(permName), squeezeId, 0);
+                    }
+                }
+                else if (axis == 1)
+                {
+                    int order[] = {0, 2, 3, 1};  // From OpenCV's NCHW to NHWC.
+                    Pin inpId = parsePin(layer.input(0));
+                    addPermuteLayer(order, name + "/nhwc", inpId);
+
+                    layerParams.set("pool", type == "Mean" ? "ave" : "sum");
+                    layerParams.set("kernel_h", 1);
+                    layerParams.set("global_pooling_w", true);
+                    int id = dstNet.addLayer(name, "Pooling", layerParams);
+                    layer_id[name] = id;
+                    connect(layer_id, dstNet, inpId, id, 0);
+
+                    if (!keepDims)
+                    {
+                        LayerParams squeezeLp;
+                        std::string squeezeName = name + "/squeeze";
+                        CV_Assert(layer_id.find(squeezeName) == layer_id.end());
+                        int channel_id = 3; // TF NHWC layout
+                        squeezeLp.set("axis", channel_id - 1);
+                        squeezeLp.set("end_axis", channel_id);
+                        int squeezeId = dstNet.addLayer(squeezeName, "Flatten", squeezeLp);
+                        layer_id[squeezeName] = squeezeId;
+                        connect(layer_id, dstNet, Pin(name), squeezeId, 0);
+                    }
+                    else
+                    {
+                        int order[] = {0, 3, 1, 2};  // From NHWC to OpenCV's NCHW.
+                        Pin inpId = parsePin(name);
+                        addPermuteLayer(order, name + "/nchw", inpId);
                     }
                 }
             } else {

--- a/modules/dnn/test/test_tf_importer.cpp
+++ b/modules/dnn/test/test_tf_importer.cpp
@@ -135,6 +135,16 @@ TEST_P(Test_TensorFlow_layers, reduce_sum)
     runTensorFlowNet("sum_pool_by_axis");
 }
 
+TEST_P(Test_TensorFlow_layers, reduce_sum_channel)
+{
+    runTensorFlowNet("reduce_sum_channel");
+}
+
+TEST_P(Test_TensorFlow_layers, reduce_sum_channel_keep_dims)
+{
+    runTensorFlowNet("reduce_sum_channel", false, 0.0, 0.0, false, "_keep_dims");
+}
+
 TEST_P(Test_TensorFlow_layers, conv_single_conv)
 {
     runTensorFlowNet("single_conv");

